apps/voting/contracts/Voting.sol

```diff
@@ -394,17 +496,85 @@ contract Voting is IForwarder, AragonApp {
     }
 
     /**
-    * @dev Internal function to check if a vote is still open
+    * @dev Internal function to check if a representative can vote on behalf of a voter on a certain vote. It assumes the queried vote exists.
+    * @return True if the given representative can vote on behalf of a voter on a certain vote, false otherwise
+    */
+    function _canVoteOnBehalfOf(Vote storage vote_, address _voter, address _representative) internal view returns (bool) {
+        return _canVote(vote_, _voter) &&
+               _isRepresentativeOf(_voter, _representative) &&
+               _hasNotVotedYet(vote_, _voter) &&
+               !_withinOverruleWindow(vote_);
+    }
+
+    /**
+    * @dev Internal function to check if a representative is allowed to vote on behalf of a voter
+    * @return True if the given representative is allowed by a certain voter, false otherwise
+    */
+    function _isRepresentativeOf(address _voter, address _representative) internal view returns (bool) {
+        return representatives[_voter] == _representative;
+    }
+
+    /**
+    * @dev Internal function to check if a voter has already voted on a certain vote. It assumes the queried vote exists.
+    * @return True if the given voter has not voted on the requested vote, false otherwise
+    */
+    function _hasNotVotedYet(Vote storage vote_, address _voter) internal view returns (bool) {
+        return _voteIssuer(vote_, _voter) != _voter;
+    }
+
+    /**
+    * @dev Internal function to check if a vote is still open. It assumes the queried vote exists.
     * @return True if the given vote is open, false otherwise
     */
     function _isVoteOpen(Vote storage vote_) internal view returns (bool) {
-        return getTimestamp64() < vote_.startDate.add(voteTime) && !vote_.executed;
+        return getTimestamp64() < _voteEndDate(vote_) && !vote_.executed;
+    }
+
+    /**
+    * @dev Internal function to check if a vote is within its overrule window. It assumes the queried vote exists.
+    * @return True if the given vote is within its overrule window, false otherwise
+    */
+    function _withinOverruleWindow(Vote storage vote_) internal view returns (bool) {
+        return getTimestamp64() >= _voteEndDate(vote_).sub(overruleWindow);
+    }
+
+    /**
+    * @dev Internal function to calculate the end date of a vote. It assumes the queried vote exists.
+    */
+    function _voteEndDate(Vote storage vote_) internal view returns (uint64) {
+        return vote_.startDate.add(voteTime);
+    }
+
+    /**
+    * @dev Internal function to get the state of a voter. It assumes the queried vote exists.
+    */
+    function _voterState(Vote storage vote_, address _voter) internal view returns (VoterState) {
+        return vote_.voters[_voter];
+    }
+
+    /**
+    * @dev Internal function to get the issuer of a vote. It assumes the queried vote exists.
+    */
+    function _voteIssuer(Vote storage vote_, address _voter) internal view returns (address) {
```

Comment:
Nitpick on the variable names, since it's a bit confusing to read some of these new internal functions. We use a prefixed `_` (e.g. `_vote`) for the argument name and a suffix `_` in variable names (e.g. `vote_`) only if it would otherwise shadow a storage variable.

Reply:
Sure, do you mean I'm not following that convention here?

Reply:
Yes, the `vote_` variable should be `_vote` since it's an argument.

Reply:
Note that we are always using `vote_` for vote storage pointers cause we would shadow the `_vote` function otherwise

Reply:
:(

Maybe we can find a different name (e.g. `_voteAsPrinciple()`)?